query-engine/prisma-models/src/field_selection.rs

```diff
@@ -0,0 +1,319 @@
+    }
+
+    /// Checks if `self` only contains scalar field selections and if so, returns them all in a list.
+    /// If any other selection is contained, returns `None`.
+    pub fn as_scalar_fields(&self) -> Option<Vec<ScalarFieldRef>> {
+        let scalar_fields = self
+            .selections()
+            .filter_map(|selection| match selection {
+                SelectedField::Scalar(sf) => Some(sf.clone()),
+                SelectedField::Composite(_) => None,
+            })
+            .collect_vec();
+
+        if scalar_fields.len() == self.selections.len() {
+            Some(scalar_fields)
+        } else {
+            None
+        }
+    }
+
+    /// Inserts this `FieldSelection`s selections into the given `SelectionResult`.
+    /// Assumes caller knows that the exchange can be done, but still errors if lengths mismatch.
+    /// Additionally performs a type coercion based on the source and destination field types.
+    /// Resistance is futile.
+    pub fn assimilate(&self, values: SelectionResult) -> crate::Result<SelectionResult> {
+        if self.selections.len() != values.len() {
+            Err(DomainError::ConversionFailure(
+                "field values".to_owned(),
+                "assimilated field values".to_owned(),
+            ))
+        } else {
+            let pairs = values
+                .pairs
+                .into_iter()
+                .zip(self.selections())
+                .map(|((_, value), new_selection)| {
+                    let value = new_selection.coerce_value(value)?;
+                    Ok((new_selection.clone(), value))
+                })
+                .collect::<crate::Result<Vec<_>>>()?;
+
+            Ok(SelectionResult::new(pairs))
+        }
+    }
+
+    /// Checks if a given `SelectionResult` belongs to this `FieldSelection`.
+    pub fn matches(&self, result: &SelectionResult) -> bool {
+        result.pairs.iter().all(|(rt, _)| self.selections.contains(rt))
+    }
+
+    /// Merges all given `FieldSelection` a set union of all.
+    /// Each selection is contained exactly once, with the first
+    /// occurrence of the first field in order from left to right
+    /// is retained.
+    ///
+    /// /!\ Important assumption: All selections are on the same model.
+    pub fn union(selections: Vec<Self>) -> Self {
+        let chained = selections.into_iter().flatten();
+
+        FieldSelection {
+            selections: chained.unique().collect(),
+        }
+    }
+
+    /// Consumes both `FieldSelection`s to create a new one that contains a union
+    /// of both. Each selection is contained exactly once, with the first
+    /// occurrence of the first field in order from left (`self`) to right (`other`)
+    /// is retained. Assumes that both selections reason over the same model.
+    pub fn merge(self, other: FieldSelection) -> FieldSelection {
+        let selections = self.selections.into_iter().chain(other.selections).unique().collect();
```

Comment:
Do you need to flatten the selections here?

Reply:
No, because of the `IntoIterator` of `FieldSelection` I can chain the two iterators together into one iterator of the same elements.